### ISSUE TITLE
When taking a snapshot of a shared sstable only one shard should create_links

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#6337]
Scylla version 3b31acfa80338314717d438f7491c499dccf167f

Currently, when taking a snapshot, it looks like multiple shards call create_links on the same source sstable to the same destination snapshot directory in parallel.

Although this is allowed it is sub-optimal and we can do with only one of the shards interested in the sstable linking it to the snapshot dir.

/cc @raphaelsc

**Comment by raphaelsc** [comment]
/cc @glommer 

After off-strategy work, table's sstable set will only contain unshared SSTables, so most of these problems related to redundant work due to shared SSTables will be gone.

As you know, we also have metric related issues with shared SSTables data being accounted more than once, at every shard they belong to.

**Comment by bhalevy** [comment]
@slivne it'd best if @raphaelsc or someone else from @glommer's team will own this issue as it @raphaelsc is already familiar with the issue and the implementation.  @glommer let me know if that is a problem and I'll take it back...

**Comment by raphaelsc** [comment]
Turns out we'll needlessly try to create links at all shards a shared SSTable belongs to, but only one of the owner shards will succeed in creating the links, while all the others will fail with EEXIST but they're able to go through it by swallowing the EEXIST exception. I think this is ugly, and we could improve this situation by only trying link creation at one of the owner shards. This can be fixed with the following patch:

```
diff --git a/table.cc b/table.cc
index ea36ee3e7..cef7a5624 100644
--- a/table.cc
+++ b/table.cc
@@ -1691,6 +1691,11 @@ future<> table::snapshot(sstring name) {
             auto jsondir = _config.datadir + "/snapshots/" + name;
             return io_check([jsondir] { return recursive_touch_directory(jsondir); }).then([this, name, jsondir, &tables] {
                 return parallel_for_each(tables, [name](sstables::shared_sstable sstable) {
+                    // Only one shard that own the sstable will submit it for snapshot to avoid race condition.
+                    if (*boost::min_element(sstable->get_shards_for_this_sstable()) != this_shard_id()) {
+                        return make_ready_future<>();
+                    }
+
                     auto dir = sstable->get_dir() + "/snapshots/" + name;
                     return io_check([dir] { return recursive_touch_directory(dir); }).then([sstable, dir] {
                         return sstable->create_links(dir).then_wrapped([] (future<> f) {
-- 
2.26.2
```

@glommer Given that this is not an actual correctness issue, but rather only a enhancement, we can set this issue as a low priority item.

**Comment by raphaelsc** [comment]
Anyway, after off-strategy work, we will don't even need to do this, so once off-strategy work is completed, we can just close this issue.

**Comment by glommer** [comment]
Thanks for the analysis Raphael.
This doesn't depend on the whole off-strategy (just nitpicking), but reshard-at-boot.

This won't take long

**Comment by bhalevy** [comment]
On Tue, May 19, 2020, 02:11 Raphael Carvalho <notifications@github.com>
wrote:

> Turns out we'll needlessly try to create links at all shards a shared
> SSTable belongs to, but only one of the owner shards will succeed in
> creating the links, while all the others will fail with EEXIST but they're
> able to go through it by swallowing the EEXIST exception. I think this is
> ugly, and we could improve this situation by only trying link creation at
> one of the owner shards. This can be fixed with the following patch:
>
> diff --git a/table.cc b/table.cc
> index ea36ee3e7..cef7a5624 100644
> --- a/table.cc
> +++ b/table.cc
> @@ -1691,6 +1691,11 @@ future<> table::snapshot(sstring name) {
>              auto jsondir = _config.datadir + "/snapshots/" + name;
>              return io_check([jsondir] { return recursive_touch_directory(jsondir); }).then([this, name, jsondir, &tables] {
>                  return parallel_for_each(tables, [name](sstables::shared_sstable sstable) {
> +                    // Only one shard that own the sstable will submit it for snapshot to avoid race condition.
> +                    if (*boost::min_element(sstable->get_shards_for_this_sstable()) != this_shard_id()) {
> +                        return make_ready_future<>();
> +                    }
> +
>
> Can this cause load imbalance towards lower ranking shards?

Can we use the sstable generation to determine the owning shard or is it
not guaranteed to be listed here?

**Comment by bhalevy** [comment]
@raphaelsc can we close the issue now that off-strategy work is in?

**Comment by raphaelsc** [comment]
sure, closing it... thanks!
